MSC3291: Muting in VoIP calls (#3291)
matrix-org/matrix-spec-proposals: proposals/3291-muting.md

This MSC also adds a new call event `m.call.sdp_stream_metadata_changed`, which
has the common VoIP fields as specified in
[MSC2746](https://github.com/matrix-org/matrix-doc/pull/2746) (`version`,
`call_id`, `party_id`) and a `sdp_stream_metadata` object which is the same
thing as `sdp_stream_metadata`  in `m.call.negotiate`, `m.call.invite` and
`m.call.answer`. The client sends this event the when `sdp_stream_metadata` has
changed but no negotiation is required (e.g. the user mutes their
camera/microphone).

If a client sends an event with some of the fields missing the previous state
should be assumed. If there is no previous state all tracks should be assumed as
*not* muted. If a track is missing from a stream that track should be considered
muted.

### Example

```JSON
{
    "type": "m.call.sdp_stream_metadata_changed",
    "room_id": "!roomId",
    "content": {
        "call_id": "1414213562373095",
        "party_id": "1732050807568877",
        "sdp_stream_metadata": {
            "2311546231": {
                "purpose": "m.usermedia",
                "audio_muted:": false,
                "video_muted": true,
            }
        },

>         "version": "1",

Reviewer: Should this be a string or an integer? I noticed that existing call events seem to use the latter.

Author: https://github.com/matrix-org/matrix-doc/blob/dbkr/msc2746/proposals/2746-reliable-voip.md#change-the-version-field-in-all-voip-events-to-1 seems to suggest both are possible

Reviewer 2: They're strings as of MSC2746

>     },

Reviewer: ```suggestion
    }
```

}
```

## Potential issues

When the user mutes their camera, the client will meaningless which will waste

> bandwidth.

Reviewer: I think you may have missed a word in this sentence?

Author: I indeed did :sweat_smile:

## Alternatives

### Only disabling the corresponding track

This is the solution that some clients (e.g. Element) use at the moment. While
this is almost instantaneous, it doesn't allow the other side to know the
opponent's mute state. This leads to the opponent showing a black screen for a
muted video track and not doing anything for a muted audio track which is bad
for UX.

### Setting the corresponding track as `recvonly`/`inactive`

While this would be beneficial for low bandwidth connections, it takes time. The

> delay might be acceptable for video but isn't for audio (with which your would

Reviewer: ```suggestion
delay might be acceptable for video but isn't for audio (with which you would
```

assume an instantaneous mute state change). This is also problematic since there
could be a confusion with holding (as defined in
[MSC2746](https://github.com/matrix-org/matrix-doc/pull/2746)).

### Using a separate event for muting

While this might feel clearer initially, it doesn't have much real benefit. The
mute state is in fact a meta information about the stream and using
`sdp_stream_metadata` is also more flexible for cases where the user joins a
call already muted. It is also more flexible in general and would be useful if
we ever decided to do what is described in the next section.

### A combination of disabling tracks, `sdp_stream_metadata` and SDP

An option would be using the current method in combination with setting the
corresponding track as `recvonly`/`inactive`. Along with this clients would need

> to set the mute state in `sdp_stream_metadata` to avoid conflicts with holding.

Reviewer: It may be useful to briefly mention how changing the track to `recvonly` relates to holding a call.

While this solution might be the most flexible solution as it would allow
clients to choose between bandwidth and a mute state change delay for each
track, it would be harder to implement and feels generally disjointed.

## Security considerations

> None that I can think of.

Reviewer: We may want to consider who can set the mute state for which user? Power level permissions may also prevent users from muting themselves, which seems wrong.

Author: This isn't really the concern of this MSC. That will become an issue when group calls over Matrix are a thing. This MSC focuses on muting only yourself which you always should have the permission to do

Reviewer: What about the ability to mute other users in the call though? Is that presumed to be a feature or a bug?

Author: That is presumed to not be possible - the event's target is always the sender, so you can't mute anyone besides yourself 

(muting other users is of course a thing that we will want in the future but that would require a separate MSC)

Reviewer: Gotcha - thanks for clarifying. Does that mean it won't ever be possible to mute other participants? Other conference software allows one to if they hold moderation status or somesuch.

Author: Muting other participants might be possible in some way but will require an MSC of its own.

Author: @anoadragon453, from what I gather I can mark this as resolved. Shout if you disagree

mute state is in fact a meta information about the stream and using
`sdp_stream_metadata` is also more flexible for cases where the user joins a
call already muted. It is also more flexible in general and would be useful if
we ever decided to do what is described in the next section.

### A combination of disabling tracks, `sdp_stream_metadata` and SDP

An option would be using the current method in combination with setting the
corresponding track as `recvonly`/`inactive`. Along with this clients would need
to set the mute state in `sdp_stream_metadata` to avoid conflicts with holding
(as defined in [MSC2746](https://github.com/matrix-org/matrix-doc/pull/2746)).
While this solution might be the most flexible solution as it would allow
clients to choose between bandwidth and a mute state change delay for each
track, it would be harder to implement and feels generally disjointed.

## Security considerations

None that I can think of.

## Unstable prefix

|Release                             |Development                                           |
|------------------------------------|------------------------------------------------------|

> |`m.call.sdp_stream_metadata_changed`|`org.matrix.msc3291.call.sdp_stream_metadata_changed` |

Author: We should also be using the prefixed `sdp_stream_metadata` from MSC3077 until that is merged

# MSC3291: Muting in VoIP calls

During VoIP calls, it is common for a user to mute their microphone/camera.
Ideally, the other side should be able to see that the opponent's camera is
muted, so that it could reflect this in the UI (e.g. show the user's avatar
instead of their camera feed). We would also want the changes in the mutes state
to be quick.

Using pure WebRTC there are two ways to do muting and both have their issues:

+ Disabling the corresponding track
+ Setting the corresponding track as `recvonly`/`inactive`

The Alternatives section describes the issues with using these alone.

## Proposal

This MSC proposes that clients disable muted tracks (as is almost instantaneous)
and extends the `SDPStreamMetadata` object (see

> [MSC3077](https://github.com/matrix-org/matrix-doc/pull/3077)) to allow

Reviewer 2: Ah yep, good point.

assume an instantaneous mute state change). This is also problematic since there
could be a confusion with holding (as defined in
[MSC2746](https://github.com/matrix-org/matrix-doc/pull/2746)).

### Using a separate event for muting

While this might feel clearer initially, it doesn't have much real benefit. The
mute state is in fact a meta information about the stream and using
`sdp_stream_metadata` is also more flexible for cases where the user joins a
call already muted. It is also more flexible in general and would be useful if
we ever decided to do what is described in the next section.

### A combination of disabling tracks, `sdp_stream_metadata` and SDP

An option would be using the current method in combination with setting the
corresponding track as `recvonly`/`inactive`. Along with this clients would need
to set the mute state in `sdp_stream_metadata` to avoid conflicts with holding
(as defined in [MSC2746](https://github.com/matrix-org/matrix-doc/pull/2746)).
While this solution might be the most flexible solution as it would allow
clients to choose between bandwidth and a mute state change delay for each

> track, it would be harder to implement and feels generally disjointed.

Reviewer 3: Are we sure we don't want to go for optimal bandwidth usage? I personally often turn off video to have more bandwidth available for audio when the quality is terrible. Feels unfortunate to exclude this. Also, this doesn't feel much harder to implement.

Reviewer 4: I think this should be the choice of the client. In generally not sending the video when video is off makes sense. The extra latency of starting up the stream is usually quite small and the bandwidth savings is significant which can make a huge difference for someone on a connection that struggle to send audio and video but can support audio reliably. Turning off the camera is less often used because actually getting a new stream from the camera can have very high and unpredictable latency depending on the hardware or the OS. Probably a good client will have a variety of heuristics to decide how far to shut down the camera but the key thing from the spec point of view that completely turning off the camera is possible if desired.

Author: I don't think I stand against going with both options at the same time - let the client choose. @dbkr, would you agree?

Author: FWIW, we do this now: https://github.com/matrix-org/matrix-js-sdk/pull/3028

Reviewer 2: Also, I believe Enrico recently discovered with his experiments that browsers (at least some of them) now send nothing when a video track is disabled, so the bandwidth argument is even less relevant. Turning cameras off is orthogonal and can be done whether or not you remove the video track. The negotiate mechanism still exists either way.

Author: > I believe Enrico recently discovered with his experiments that browsers (at least some of them) now send nothing when a video track is disabled, so the bandwidth argument is even less relevant

Is this documented anywhere?

Reviewer: I'm going to assume not and resolve this thread. Please re-open if you believe otherwise.

Ideally, the other side should be able to see that the opponent's camera is
muted, so that it could reflect this in the UI (e.g. show the user's avatar
instead of their camera feed). We would also want the changes in the mutes state
to be quick.

Using pure WebRTC there are two ways to do muting and both have their issues:

+ Disabling the corresponding track
+ Setting the corresponding track as `recvonly`/`inactive`

The Alternatives section describes the issues with using these alone.

## Proposal

This MSC proposes extending the `SDPStreamMetadata` object (see
[MSC3077](https://github.com/matrix-org/matrix-doc/pull/3077)) to allow
indicating the mute state to the other side using the following fields:

+ `audio_muted` - a boolean indicating the current audio mute state
+ `video_muted` - a boolean indicating the current video mute state

This MSC also adds a new call event `m.call.sdp_stream_metadata_changed`, which
has the common VoIP fields as specified in
[MSC2746](https://github.com/matrix-org/matrix-doc/pull/2746) (`version`,
`call_id`, `party_id`) and a `sdp_stream_metadata` object which is the same
thing as `sdp_stream_metadata` in `m.call.negotiate`, `m.call.invite` and

> `m.call.answer`. The client sends this event the when `sdp_stream_metadata` has

Reviewer: ```suggestion
`m.call.answer`. The client sends this event when the `sdp_stream_metadata` has
```

changed but no negotiation is required (e.g. the user mutes their
camera/microphone).

All tracks should be assumed unmuted unless specified otherwise.

It is suggested `video_muted` (but not required) should determine if a user sees
video but `audio_muted` only determines if a user is shown a mute indication
rather than actually muting the audio track on the receiving side. This is

> suggested because if the track is muted using the WebRTC `disabled` property,

Reviewer: I tried to look for the `disabled` property on MDN, but only found [`MediaStreamTrack.enabled`](https://developer.mozilla.org/en-US/docs/Web/API/MediaStreamTrack/enabled), which can be set to `false`. Is that what you're referencing?

> It is suggested `video_muted` (but not required) should determine if a user sees
> video but `audio_muted` only determines if a user is shown a mute indication
> rather than actually muting the audio track on the receiving side. This is
> suggested because if the track is muted using the WebRTC `disabled` property,
> there can be delay between the audio track getting unmuted and the client
> receiving the `m.call.sdp_stream_metadata_changed` event.

Reviewer: As someone who isn't familiar with WebRTC development, this took a few reads to actually parse what it was suggesting I do as a (theoretical) client developer. The following is my attempt to make it a bit clearer:

```suggestion
Clients are recommended to not set the WebRTC `disabled` property on their local audio
stream when a stream has the `audio_muted` field set to `false`. This is because when
the other user unmutes themselves, there may be a slight delay between their client
sending audio and the `m.call.sdp_stream_metadata_changed` event arriving. If
`disabled` is set to `true`, then any audio sent in between those two events will not
be heard. The other user will still stop transmitting audio once they mute on their
side, so no audio is sent without the user's knowledge.

The same suggestion does not apply to `video_muted`.
```

Also... why does the same not apply to `video_muted`? Wouldn't you then potentially miss a little bit of video stream?

### Example

```JSON
{
    "type": "m.call.sdp_stream_metadata_changed",
    "room_id": "!roomId",
    "content": {
        "version": "1",
        "call_id": "1414213562373095",
        "party_id": "1732050807568877",
        "sdp_stream_metadata": {
            "2311546231": {
                "purpose": "m.usermedia",
                "audio_muted:": true,
                "video_muted": true
            }
        }
    }
}
```

This event indicates that both audio and video are muted. It is suggested the
video track of stream `2311546231` should be hidden in the UI (probably replaced
by an avatar). It also suggests the UI should show an indication that the audio

> track is muted but the client should not mute the audio on the receiving side.

Reviewer: ```suggestion
track is muted but, as above, the client should not mute the audio on the
receiving side.
```

Author: What do you mean by the _as above_?

Reviewer: Ah, it's referencing the paragraph above which talks about not locally muting an incoming audio stream. When I initially read this, I thought "why should the client should not mute the audio on the receiving side?" before I remembered the reasoning above.

"as above" could be replaced with "as explained earlier" to be a bit clearer.

This suggestion is mostly personal preference though - feel free to yay/nay it :)

## Potential issues

When the user mutes their camera, the client will keep sending meaningless data

> which will waste bandwidth.

Reviewer: What, why? Do you mean only for a short period, or the whole time a user's camera is muted? Currently it sounds like the latter.

Author: I am not sure, if this is still the case but some browsers used to send black pixels (meaningless data) when the track got disabled

Reviewer: Ah OK. Best to mention that this isn't a choice by clients then.

```suggestion
When the user mutes their camera, some browsers may keep sending meaningless data
which will waste bandwidth.
```

## Alternatives

### Only disabling the corresponding track

> This is the solution that some clients (e.g. Element) use at the moment. While

Reviewer: I assume Element doesn't do this any more now that it implemented this MSC? :yum:

`sdp_stream_metadata` is also more flexible for cases where the user joins a
call already muted. It is also more flexible in general and would be useful if
we ever decided to do what is described in the next section.

### A combination of disabling tracks, `sdp_stream_metadata` and SDP

An option would be using the current method in combination with setting the
corresponding track as `recvonly`/`inactive`. Along with this clients would need
to set the mute state in `sdp_stream_metadata` to avoid conflicts with holding
(as defined in [MSC2746](https://github.com/matrix-org/matrix-doc/pull/2746)).
While this solution might be the most flexible solution as it would allow
clients to choose between bandwidth and a mute state change delay for each
track, it would be harder to implement and feels generally disjointed.

## Security considerations

None that I can think of.

## Dependencies

+ [MSC3077](https://github.com/matrix-org/matrix-doc/pull/3077)

## Unstable prefix

|Release                             |Development                                  |
|------------------------------------|---------------------------------------------|

> |`m.call.sdp_stream_metadata_changed`|`org.matrix.call.sdp_stream_metadata_changed`|

Author: Argh, the `js-sdk` actually uses this... Is it fine, if we leave it this way?

> |`m.call.sdp_stream_metadata_changed`|`org.matrix.call.sdp_stream_metadata_changed`|

Reviewer: Generally we do not want MSCs to pollute the `org.matrix.*` namespace to avoid collisions. But given that a) if one does, we remain fine as long as we do not let another MSC to do so and b) changing the implementation in favour of arbitrary correctness would be much more of a headache for everyone involved, I'd say no, you don't need to update the implementation. Especially as the unstable prefix will eventually be dropped if this MSC (hopefully) reaches acceptance soon.

Curiously though -- I did notice that this MSC *used* to have an unstable prefix of `org.matrix.msc3077` for this field, but that was dropped in https://github.com/matrix-org/matrix-spec-proposals/pull/3291/commits/f1ac61f94264af05b70dc5fb71de6238e063615f. Do you happen to remember why that happened, so we may prevent it in future?

> |`m.call.sdp_stream_metadata_changed`|`org.matrix.call.sdp_stream_metadata_changed`|

Author: > Generally we do not want MSCs to pollute the `org.matrix.*` namespace to avoid collisions. But given that a) if one does, we remain fine as long as we do not let another MSC to do so and b) changing the implementation in favour of arbitrary correctness would be much more of a headache for everyone involved, I'd say no, you don't need to update the implementation. Especially as the unstable prefix will eventually be dropped if this MSC (hopefully) reaches acceptance soon.

Good to hear... :sweat_smile: 

> Curiously though -- I did notice that this MSC _used_ to have an unstable prefix of `org.matrix.msc3077` for this field, but that was dropped in [f1ac61f](https://github.com/matrix-org/matrix-spec-proposals/commit/f1ac61f94264af05b70dc5fb71de6238e063615f). Do you happen to remember why that happened, so we may prevent it in future?

Hah, if only I knew that...

# MSC3291: Muting in VoIP calls

During VoIP calls, it is common for a user to mute their microphone/camera.
Ideally, the other side should be able to see that the opponent's camera is
muted, so that it could reflect this in the UI (e.g. show the user's avatar
instead of their camera feed). We would also want the changes in the mutes state
to be quick.

Using pure WebRTC there are two ways to do muting and both have their issues:

+ Disabling the corresponding track
+ Setting the corresponding track as `recvonly`/`inactive`

The Alternatives section describes the issues with using these alone.

## Proposal

> This MSC proposes extending the `SDPStreamMetadata` object (see

Reviewer 5: MSC3077 doesn't actually define a thing called `SDPStreamMetadata`

```suggestion
This MSC proposes extending the `sdp_stream_metadata` object (see
```

# MSC3291: Muting in VoIP calls

During VoIP calls, it is common for a user to mute their microphone/camera.
Ideally, the other side should be able to see that the opponent's camera is
muted, so that it could reflect this in the UI (e.g. show the user's avatar
instead of their camera feed). We would also want the changes in the mutes state
to be quick.

Using pure WebRTC there are two ways to do muting and both have their issues:

+ Disabling the corresponding track - there will be no way to know the track is
  muted from the opponent's perspective. On the other hand, this is almost
  instantaneous.

> + Setting the corresponding track as `recvonly`/`inactive` - this is problematic

Reviewer 6: So the normal IETF SIP way to indicate muted media is to set the track as recvonly by sending an updated SDP offer. Does this actually need a renegotiation though?  It would if it was SIP, but in Matrix, could we implicitly assume the same SDP answer will apply?

Pros:
 * It maps easily to other SDP-speaking protocols (e.g. SIP) and doesn't reinvent the wheel
 * It avoids mixing together SDP-described media with adhoc JSON-described media (as unpleasant as SDP is, i'm worried about ending up with a session description protocol which is half SDP and half JSON)
 * It's no more events than sending a `m.call.sdp_stream_metadata_changed` event
Cons:
 * You waste bandwidth sending all the rest of the SDP through when you've just changed a single line.
 * It'd have to be designed to not clash with call hold behaviour.

I don't feel /that/ strongly about this, but I think we should at least list this as an alternative and justify the chosen solution a bit more clearly.

Author: While this sounds interesting, I see two main issues:

+ Doing this using SDP will always be slower, afaik. This would be alright for video but for audio, you'd assume the other side can hear you right after you unmute yourself, which wouldn't happen if we did this with SDP.
+ How would this be technically done, would we fake the remote's answer?

We've discussed doing this using SDP with @dbkr and we came to the conclusion that the delay is just too big of a problem (at least for audio). We could probably try doing this using SDP for video and use the `video_muted` to differentiate from holding but it felt weird to have two ways to do muting

Author: I've added an Alternatives section which hopefully makes things a bit clearer

Reviewer 6: my point is that it should be no slower to send a new sdp offer than sending a custom json event. we don’t need to do a full handshake or negotiation with a new answer to the offer; we can reuse the one we already have from the initial setup.

Author: > my point is that it should be no slower to send a new sdp offer than sending a custom json event

Not really, the JSON event is there only to tell the other side that we are muted, _not_ to do the actual muting (that is done by disabling the track). For example, the other side doesn't have to wait for this event to start receiving audio. On the other hand, if we did this using SDP, the event has to reach the other side. This of course isn't the case for video as we expect the opponent to show the user's avatar (or whatever) when muted.

> we don’t need to do a full handshake or negotiation with a new answer to the offer; we can reuse the one we already have from the initial setup.

Can we though? We would have to manually change the direction of the tracks so that the change in direction actually takes place

<hr>

Nonetheless, this sounds interesting and we should probably try this out

Reviewer 2: My main concern with this was that because we use recvonly/sendonly to infer whether someone's on hold or not, also using it to infer mute status would make things a lot harder to distinguish between the two.

On speed, you're correct that we could make the answer implicit, but the WebRTC stack will need an answer before it would do anything, so we would have to fake one (or re-use the one from before the mute), which doesn't feel great in terms of meddling with the offer/answer semantics.

> > my point is that it should be no slower to send a new sdp offer than sending a custom json event

> Not really, the JSON event is there only to tell the other side that we are muted, not to do the actual muting (that is done by disabling the track). For example, the other side doesn't have to wait for this event to start receiving audio. On the other hand, if we did this using SDP, the event has to reach the other side. This of course isn't the case for video as we expect the opponent to show the user's avatar (or whatever) when muted.

Yep, either message would have to go via the Matrix signalling path, but yep, with the custom message it's only informational, so we could start receiving audio straight away rather than having to wait for the signalling message. This feels important in terms of being able to speak immediately after unmuting. Come to think of it, we should probably add an implicit 'unmute' when we start getting audio again (so you don't get people speaking even though they still appear muted), although quite how you'd prevent this from being racy I don't know.

Reviewer 6: hum. the mix of SDP & JSON still feels icky to me. For hold, shouldn't we be offering `inactive` (given we're neither sending nor listening to the user we've put on hold?).  And then offering `recvonly` means mute. 

I agree it's icky to also mess with the offer/answer semantics by reusing the last answer when sending a new 'offer' to indicate this state.  But then we already have this handshake when putting a call on hold.  I kinda wish we'd done this JSON approach for call hold so we didn't end up with this nasty mix of half-Matrix and half-SIP behaviour.

Given the `recvonly/inactive` thing really is a nasty hangover from SIP, i guess creeping towards expressing it with JSON is better, and perhaps we'll end up with another MSC to fix the call-hold behaviour.  So, i retract my concern.

Author: > hum. the mix of SDP & JSON still feels icky to me. For hold, shouldn't we be offering inactive (given we're neither sending nor listening to the user we've put on hold?). And then offering recvonly means mute.

For holding it should actually be `sendonly` afaik, since when putting on hold you would sometimes play a song or whatever. This results in `recvonly` on the other side. And the other side can't tell if they were put on hold or muted. 

> i guess creeping towards expressing it with JSON is better, and perhaps we'll end up with another MSC to fix the call-hold behaviour

Yeah, this at least partially makes sense to me. Though may I suggest marking this as resolved and moving the conversation about holding to https://github.com/matrix-org/matrix-doc/pull/2746 ?

Reviewer 6: > For holding it should actually be sendonly afaik, since when putting on hold you would sometimes play a song or whatever. 

good point.

Also, thinking further, `m.call.candidates` is an excellent example of us augumenting the limitations of SDP with custom JSON, so actually there's a precedent for this and i think this is the best approach.

Plus using it on eleweb develop is fab :)

Author: Thanks :)

`m.call.answer`. The client sends this event the when `sdp_stream_metadata` has
changed but no negotiation is required (e.g. the user mutes their
camera/microphone).

If a client sends an event with some of the fields missing the previous state
should be assumed. If there is no previous state all tracks should be assumed as
*not* muted. If a track is missing from a stream that track should be considered
muted.

### Example

```JSON
{
    "type": "m.call.sdp_stream_metadata_changed",
    "room_id": "!roomId",
    "content": {
        "call_id": "1414213562373095",
        "party_id": "1732050807568877",
        "sdp_stream_metadata": {
            "2311546231": {
                "purpose": "m.usermedia",
                "audio_muted:": false,
                "video_muted": true,
            }
        },
        "version": "1",
    },
}
```

## Potential issues

When the user mutes their camera, the client will keep sending black pixels
which will waste bandwidth.

## Security considerations

None that I can think of.

## Unstable prefix

|Release      |Development                                           |
|-------------|------------------------------------------------------|

> |`m.call.mute`|`org.matrix.msc3291.call.sdp_stream_metadata_changed` |

Reviewer 6: ```suggestion
|`m.call.sdp_stream_metadata_changed`|`org.matrix.msc3291.call.sdp_stream_metadata_changed` |
```

[MSC2746](https://github.com/matrix-org/matrix-doc/pull/2746) (`version`,
`call_id`, `party_id`) and a `sdp_stream_metadata` object which is the same
thing as `sdp_stream_metadata` in `m.call.negotiate`, `m.call.invite` and
`m.call.answer`. The client sends this event when the `sdp_stream_metadata` has
changed but no negotiation is required (e.g. the user mutes their
camera/microphone).

All tracks should be assumed unmuted unless specified otherwise.

Clients are recommended to not mute the audio of WebRTC tracks locally when a

> stream has the `audio_muted` field set to `false`. This is because when the

Reviewer: Err, re-reading and paging this back in... did I make a typo? We're talking about not muting a stream locally when another user mutes themselves; `audio_muted` is set to `true`.

```suggestion
incoming stream has the `audio_muted` field set to `true`. This is because when the
```

I've also clarified that this is for *incoming* streams - from the other user.

Author: Ah, yeah, you're right
